packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,548 @@
+
+export const enum ENVIRONMENTAL_MASK {
+  isGhostery = 1 << 0,
+  isUnsupportedPlatform = 1 << 1,
+  isManifestV3 = 1 << 2,
+  isMobile = 1 << 3,
+
+  // Browser specs
+  isBrowserChromium = 1 << 4,
+  isBrowserFirefox = 1 << 5,
+  isBrowserSafari = 1 << 6,
+  isBrowserOpera = 1 << 7,
+
+  // Capabilities
+  hasHtmlFilteringCapability = 1 << 8,
+  hasUserStylesheetCapability = 1 << 9,
+  // RESERVE = 1 << {10...14}
+
+  // Misc
+  False = 1 << 15,
+}
+
+export const enum PREPROCESSOR_UTIL_MASK {
+  isNegated = 1 << 30,
+  isContinuedWithLogicalAndOperator = 1 << 31,
+}
+
+export class Env {
+  public readonly mask: number;
+
+  public readonly isGhostery: boolean;
+  public readonly isUnsupportedPlatform: boolean;
+  public readonly isManifestV3: boolean;
+  public readonly isMobile: boolean;
+  public readonly isBrowserChromium: boolean;
+  public readonly isBrowserFirefox: boolean;
+  public readonly isBrowserSafari: boolean;
+  public readonly isBrowserOpera: boolean;
+  public readonly hasHtmlFilteringCapability: boolean;
+  public readonly hasUserStylesheetCapability: boolean;
+  public readonly False: boolean;
+
+  constructor({
+    isGhostery = true,
+    isUnsupportedPlatform = false,
+    isManifestV3 = true,
+    isMobile = true,
+    isBrowserChromium = true,
+    isBrowserFirefox = true,
+    isBrowserSafari = true,
+    isBrowserOpera = true,
+    hasHtmlFilteringCapability = true,
+    hasUserStylesheetCapability = true,
+    False = false,
+  }: Partial<Env> = {}) {
+    this.isGhostery = isGhostery;
+    this.isUnsupportedPlatform = isUnsupportedPlatform;
+    this.isManifestV3 = isManifestV3;
+    this.isMobile = isMobile;
+    this.isBrowserChromium = isBrowserChromium;
+    this.isBrowserFirefox = isBrowserFirefox;
+    this.isBrowserSafari = isBrowserSafari;
+    this.isBrowserOpera = isBrowserOpera;
+    this.hasHtmlFilteringCapability = hasHtmlFilteringCapability;
+    this.hasUserStylesheetCapability = hasUserStylesheetCapability;
+    this.False = False;
+
+    let mask = 0;
+
+    if (this.isGhostery) mask = setBit(mask, ENVIRONMENTAL_MASK.isGhostery);
+    if (this.isUnsupportedPlatform) mask = setBit(mask, ENVIRONMENTAL_MASK.isUnsupportedPlatform);
+    if (this.isManifestV3) mask = setBit(mask, ENVIRONMENTAL_MASK.isManifestV3);
+    if (this.isMobile) mask = setBit(mask, ENVIRONMENTAL_MASK.isMobile);
+    if (this.isBrowserChromium) mask = setBit(mask, ENVIRONMENTAL_MASK.isBrowserChromium);
+    if (this.isBrowserFirefox) mask = setBit(mask, ENVIRONMENTAL_MASK.isBrowserFirefox);
+    if (this.isBrowserSafari) mask = setBit(mask, ENVIRONMENTAL_MASK.isBrowserSafari);
+    if (this.isBrowserOpera) mask = setBit(mask, ENVIRONMENTAL_MASK.isBrowserOpera);
+    if (this.hasHtmlFilteringCapability)
+      mask = setBit(mask, ENVIRONMENTAL_MASK.hasHtmlFilteringCapability);
+    if (this.hasUserStylesheetCapability)
+      mask = setBit(mask, ENVIRONMENTAL_MASK.hasUserStylesheetCapability);
+
+    this.mask = mask;
+  }
+}
```

Comment:
I am not sure if it's worth even having a hard-coded list of supported processors. My understanding is that the conditions are simple boolean expressions using constants. These constants can be arbitrary identifiers that can be defined or not and as far as the implementation is concerned I think it would be simpler to just keep the expression as is during engine initialization.

For example:
```
!#if foo and bar
filter1
filter2
filter3
#!endif
```

Would create a "preprocessor block" (stored in some datastructure, let's call it `PreprocessorBlock`) with following content:
```
PreprocessorBlock { condition: "foo and bar", filters: [filter1, filter2, filter3] }
```

(The list of filters can even be just the IDs of the filters, it does not have the be the actual instance of NetworkFilter or CosmeticFilter).

Then in my opinion, we can have the FiltersEngine instance hold a list of these preprocessor blocks:
```
FiltersEngine { preprocessors: PreprocessorBlock[], ... }
```

Whenever we initialize an instance of FiltersEngine (or using a method to update the env; but I assume we need to do it either only once, or very rarely), we can pass a custom map of env variables of arbitrary names associated with boolean values (or bits). For example:
```
env = {
  isGhostery: true,
  FooBar: true,
  uBlockOriginSomeNewVariableWeNeverSawBefore: true,
}
```

The labels don't have to be pre-defined, and don't have to have been seen before (which means filter lists can adopt new ones without us having to update the adblocker core library).

At that point we iterate on the preprocessor blocks one by one, look at their condition, parse it on the fly (this should be very fast) and evaluate based on the env above. Two scenarios:
1. The condition evaluates to `true`, in this case nothing to be done (the filters should, IMO, all be part of the FiltersEngine, regardless of preprocessor blocks)
2. The condition evaluates to `false`, in this case we add the filter IDs to a `Set<number>` that is part of FiltersEngine. 

At this point, the `Set<number>` of IDs is all we need at runtime to respect the preprocessors (until the env is updated). While matching, we apply the normal logic to find filters/exceptions/etc. but then whenever a matching filter is found, we check our `Set<number>` and if the filter's ID is part of it, we ignore it. This `Set<number>` can be passed as a reference (optionally) to the methods `match`, `matchAll`, `getCosmeticsFilters`, `getHtmlRules`, etc. of the network and cosmetics buckets instances (no need to store the reference).

Let me know if I missed anything but with this design, I think we can simplify the implementation from this PR:
1. Only FiltersEngine need to store more data
2. No need to change anything in NetworkFilter/CosmeticFilter and their buckets classes (except adding an extra method argument and a lookup on match)
3. No need to define bitmasks for anything, we evaluate the condition from its string form on the fly.

Reply:
The intention of hardcoded labels were because uBO treated the invalid labels as invalid preprocessor blocks at the first hand. I think your approach here is way more flexible and I'm willing to change my implementation here.

However, I think not using bitmask to evaluate env may cause performance shock at the moment unless we drop the support of changing environment on runtime. This means we'll only able to set the environment on the engine initialisation time.

Reply:
@remusao to be clear. Should we keep an info on a filter level to remember if it is affected by the preprocessor? Checking every matched filter againt the `Set<number>` may affect performance on a scale.

Reply:
I would be surprised if there is significant overhead with a Set lookup. Let’s measure it and decide but I am pretty sure it’s negligible.